File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        preOrderTravel(root, 0, list);
        return list;
    }
    
    private void preOrderTravel(TreeNode node, int level, List<Integer> list) {
        if (node == null) {
            return;
        }
        
        if (level == list.size()) {
            list.add(node.val);
        }
        
        preOrderTravel(node.right, level + 1, list);
        preOrderTravel(node.left, level + 1, list);
    }
}

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (root == nullptr) {
            return {};
        }
        
        std::queue<TreeNode*> q_level, q_next;
        q_level.push(root);
        std::vector<int> res;
        
        while(!q_level.empty() || !q_next.empty()) {
            TreeNode* node = q_level.front();
            q_level.pop();
            
            if (node->left) {
                q_next.push(node->left);
            }
            if (node->right) {
                q_next.push(node->right);
            }
            
            if (q_level.empty()) {
                q_level = std::move(q_next);
                res.push_back(node->val);
            }
        }
        
        return res;
    }
};